app/models/reservation.rb

```diff
@@ -17,17 +17,23 @@ class Reservation < ApplicationRecord # :nodoc:
   validates :checkin_at, presence: true
   validates :checkout_at, presence: true
   validates :book_code, presence: true, uniqueness: true
-
-  # Callbacks
-  before_create :assigned_room, if: -> { room.unavailable? }
+  validate :room_must_be_available
 
   private
 
-  def assigned_room
+  def room_must_be_available # rubocop:disable Metrics/AbcSize
+    return if room.blank? || checkin_at.blank? || checkout_at.blank?
+
+    overlapping_reservations = Reservation.where(room_id:)
+                                          .where.not(id:)
+                                          .where('checkin_at < ? AND checkout_at > ?', checkout_at, checkin_at)
+
+    errors.add(:base, 'Room is not available for the selected dates') if overlapping_reservations.exists?
+
     room = Room.find_by(status: 'available', category: self.room.category)
 
-    errors.add(:status, message: 'room is not available anymore') if room.nil?
-    # binding.pry
+    errors.add(:status, message: "There's no room with available status, select another category") if room.nil?
```

Comment:
Rather than writing down the actual UI message, you could assign a code to it, like `:unavailable_room`, and then in the UI you display the error message in question (if you want).

Reply:
Ah I see, It's time to create proper locale files and respective texts. But I'll include in another PR. Thanks for raising it.